src/HashingLogic.ts

```diff
@@ -53,35 +120,198 @@ export const orderedStringify = (obj: {}) => {
+  const sender = ethUtil.publicToAddress(pubKey)
+  return ethUtil.bufferToHex(sender)
+}
+
+/**
+ * Given an array of hashed dataNode signatures and a hashed checksum signature, creates a new MerkleTree
+ * after padding, and sorting.
+ *
+ */
+export const getBloomMerkleTree = (
+  dataHashes: string[],
+  paddingNodes: string[],
+  checksumHash: string
+): MerkleTree => {
+  let leaves = dataHashes
+  leaves.push(checksumHash)
+  leaves = leaves.concat(paddingNodes)
   return getMerkleTreeFromLeaves(leaves)
 }
 
+/**
+ * Given an array of root hashes, sort and hash them into a checksum buffer
+ * @param {string[]} dataHashes - array of dataHashes as hex strings
+ */
+export const getChecksum = (dataHashes: string[]): Buffer => {
+  return ethUtil.toBuffer(hashMessage(JSON.stringify(dataHashes.sort())))
+}
+
+/**
+ * Given an array of root hashes, get and sign the checksum
+ * @param dataHashes - array of dataHashes as hex strings
+ * @param privKey - private key of signer
+ */
+export const signChecksum = (dataHashes: string[], privKey: Buffer): string => {
+  return signHash(getChecksum(dataHashes), privKey)
+}
+
+/**
+ * Sign a complete attestation node and return an object containing the datanode and the signature
+ * @param dataNode - Complete attestation data node
+ * @param globalRevocationLink - Hex string referencing revocation of the whole attestation
+ * @param privKey - Private key of signer
+ */
+export const getSignedDataNode = (
+  dataNode: IAttestation,
+  globalRevocationLink: string,
+  privKey: Buffer
+): IDataNode => {
+  const attestationNode: IAttestationNode = {
+    data: dataNode.data,
+    type: dataNode.type,
+    aux: dataNode.aux,
+    link: {
+      local: generateNonce(),
+      global: globalRevocationLink,
+    },
+  }
+  const attestationHash = hashAttestationNode(attestationNode)
+  const attestationSig = signHash(attestationHash, privKey)
+  return {
+    attestationNode: attestationNode,
+    signedAttestation: attestationSig,
+  }
+}
+
+/**
+ * Given the number of data nodes return an array of padding nodes
+ * @param {number} dataCount - number of data nodes in tree
+ */
+export const getPadding = (dataCount: number): string[] => {
+  if (dataCount < 1) return []
+  let i = 5
+  while (dataCount + 1 > 2 ** (i - 1)) {
+    i += 5
+  }
+  const paddingCount = 2 ** (i - 1) - (dataCount + 1)
+  return Array.apply(null, Array(paddingCount)).map(
+    (item: number, index: number) => {
+      return hashMessage(crypto.randomBytes(20))
+    }
+  )
+}
+
+/**
+ * Given attestation data and the attester's private key, construct the entire Bloom Merkle tree
+ * and return the components needed to generate proofs
+ * @param dataNodes - Complete attestation nodes
+ * @param privKey - Attester private key
+ */
+export const getSignedMerkleTreeComponents = (
+  dataNodes: IAttestation[],
+  privKey: Buffer
+): IBloomMerkleTreeComponents => {
+  const globalRevocationLink = generateNonce()
```

Comment:
Will this need to be stored locally / in our db?

Reply:
yes we should store the whole thing returned by getComponents as a JSONB